From 7f6fb2959c3c26533053ac829e6e396b25aef393 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fabiano=20Fid=C3=AAncio?= <fidencio@redhat.com>
Date: Wed, 22 Jul 2015 04:05:45 +0200
Subject: [PATCH] events: allow to remove disabled timers and handles

Trying to remove a disabled timer or handle will cause
virt_viewer_events_remove_{handle,timeout}() to return an error
rather than removing it.

Based on commit 79699d73e6e1b7218e3bd8349d176752f86128b9 from
libvirt-glib.
Original author: Christophe Fergeau <cfergeau@redhat.com>

Related to: rhbz#1243228

(cherry picked from commit 4ec75ee70cbe7560ffb83e4e84ffb4f925a41390)
---
 src/virt-viewer-events.c | 20 +++++++++-----------
 1 file changed, 9 insertions(+), 11 deletions(-)

diff --git a/src/virt-viewer-events.c b/src/virt-viewer-events.c
index fe3aaa6..44740c3 100644
--- a/src/virt-viewer-events.c
+++ b/src/virt-viewer-events.c
@@ -233,12 +233,11 @@ virt_viewer_events_remove_handle(int watch)
 
     g_debug("Remove handle %d %d", watch, data->fd);
 
-    if (!data->source)
-        goto cleanup;
-
-    g_source_remove(data->source);
-    data->source = 0;
-    data->events = 0;
+    if (data->source != 0) {
+        g_source_remove(data->source);
+        data->source = 0;
+        data->events = 0;
+    }
 
     /* since the actual watch deletion is done asynchronously, a update_handle call may
      * reschedule the watch before it's fully deleted, that's why we need to mark it as
@@ -409,11 +408,10 @@ virt_viewer_events_remove_timeout(int timer)
 
     g_debug("Remove timeout %p %d", data, timer);
 
-    if (!data->source)
-        goto cleanup;
-
-    g_source_remove(data->source);
-    data->source = 0;
+    if (data->source != 0) {
+        g_source_remove(data->source);
+        data->source = 0;
+    }
 
     /* since the actual timeout deletion is done asynchronously, a update_timeout call may
      * reschedule the timeout before it's fully deleted, that's why we need to mark it as
